core/src/cluster_info_entries_listener.rs

```diff
@@ -0,0 +1,174 @@
+                .chain(vec![pubkey].into_iter())
+                .for_each(|pubkey| {
+                    // Keep a map in case multiple gossip proofs come at once
+                    let mut chunks_per_slot: HashMap::<Slot, (Vec<DuplicateShred>, usize)> = HashMap::new();
+                    cluster_info
+                        .get_duplicate_shreds_from(&pubkey)
+                        .filter(|chunk| blockstore.get_duplicate_slot(chunk.slot).is_none()) // Filter out slots we already know are duplicate
+                        .for_each(|chunk| match chunks_per_slot.entry(chunk.slot) {
+                            Entry::Vacant(entry) => {
+                                let mut chunks = Vec::new();
+                                let num_chunks = chunk.num_chunks.into();
+                                chunks.push(chunk);
+                                entry.insert((chunks, num_chunks));
+                            }
+                            Entry::Occupied(mut entry) => {
+                                let (chunks, _) = entry.get_mut();
+                                chunks.push(chunk);
+                            }
+                        });
+
+                    chunks_per_slot.into_iter().for_each(|(slot, (chunks, num_chunks)): (Slot, (Vec<DuplicateShred>, usize))| {
+                        let leader = |slot : Slot| {
+                            poh_recorder.lock().unwrap().bank().as_ref().map(|bank| slot_leader_at(slot, bank)).flatten()
+                        };
+                        // See if we have all of the shreds needed for the proof yet
+                        match chunks.len().cmp(&num_chunks) {
+                            Equal => match duplicate_shred::into_shreds(chunks.into_iter(), leader) {
+                                Ok((shred1, shred2)) => {
+                                    if let Err(e) = blockstore.store_duplicate_slot(
+                                        slot,
+                                        shred1.payload,
+                                        shred2.payload,
+                                    ) {
+                                        error!("Unable to store duplicate slot {} in blockstore: {}", slot, e)
+                                    }
+                                    if let Err(e) = duplicate_slot_sender.send(slot) {
+                                        error!("Unable to notify replay stage of duplicate slot {} from gossip proof: {}", slot, e)
+                                    }
+                                }
+                                Err(e) => {
+                                    warn!("Unable to ingest duplicate slot proof for {}: {}", slot, e)
+                                }
+                            },
+                            Greater => error!("Duplicate slot proof for {} is corrupt, expected {} chunks but got {} chunks", slot, num_chunks, chunks.len()),
+                            _ => (),
```

Comment:
Maybe a comment here about what happens when we have insufficient chunks to reconstruct the shreds. Something along the lines of:

"We will retry when we get the missing chunks via another signal from `duplicate_shred_receiver`"